### PR TITLE
feat(ProcessProposal)!: Check for message inclusion using subtree roots

### LINKED ISSUE #382: Tracking: Non-interactive defaults
With our current message inclusion check, we are not checking the location of the messages, only that each message has a corresponding commitment in a `MsgPayForData`. This works for ensuring that each message was included, but this doesn't actually check that the commitment is actually a subtree root to one of the data availability header row or column roots.

In order to do this, we have to have block producers organize the message layout to follow the [non-interactive defaults](https://github.com/celestiaorg/celestia-specs/blob/master/src/rationale/message_block_layout.md). This will involve a refactor during the app's `PrepareProposal` call, along with adding the message [starting location to wrapped txs](https://github.com/celestiaorg/celestia-core/issues/462#issuecomment-1118738645).

Additionally, we need validators and full nodes to check that each share commitment is actually a subtree root to one of the data availability headers. This will require a refactor to the app's `ProcessProposal` call, and is blocked by the functionality described in the above paragraph.

- [x] Add `MessageStartIndex` field to the [malleated transactions](https://github.com/celestiaorg/celestia-core/blob/7e29d6560b7e2c96c60c2c5b2ae5d9e371e382d8/proto/tendermint/types/types.proto#L221-L227) struct https://github.com/celestiaorg/celestia-core/issues/462
- [x] Create a good API for checking for subtree root inclusion #381
- [x] Refactor `PrepareProposal` to organize the message layout according to the non-interactive defaults
- [x] #383

A **huge bonus** would be if we would also implement message inclusion fraud proofs, but this does not have to be completed to close this issue

### LINKED ISSUE #383: Refactor`ProcessProposal` to check that subtree roots are actually part of data root

### PR DESCRIPTION
**Pull request by evan-forbes**
This PR updates `ProcessProposal` to use subtree roots to check for message inclusion, instead of only relying on comparing the commitments. 

This is last non-interactive defaults PR :tada: 

closes #383

### CODE REVIEW

**Comment by evan-forbes** [comment]
will rebase to fix the git history

File: x/payment/types/payfordata.go [hunk 1]
```diff
@@ -150,13 +150,16 @@ func CreateCommitment(squareSize uint64, namespace, message []byte) ([]byte, err
...
+		tree := nmt.New(sha256.New())
 		for _, leaf := range set {
 			nsLeaf := append(make([]byte, 0), append(namespace, leaf...)...)
-			// here we hardcode pushing as axis 0 cell 0 because we never want
-			// to add the parity namespace to our shares when we create roots.
-			tree.Push(nsLeaf, rsmt2d.SquareIndex{Axis: 0, Cell: 0})
+			err := tree.Push(nsLeaf)
+			if err != nil {
+				return nil, err
+			}
```
Comment (evan-forbes) [inline x/payment/types/payfordata.go:160]: had to revert this change from #618 to get the commitments to match, which was really weird. I'm still unsure why this is and requires further debugging. We are using the nmt wrapper to create row and col roots, so it only makes sense to use them when we create commitment so that they match.

File: pkg/shares/utils.go [hunk 2]
```diff
@@ -1,16 +1,17 @@
...
 	core "github.com/tendermint/tendermint/proto/tendermint/types"
 	coretypes "github.com/tendermint/tendermint/types"
 )
 
 // DelimLen calculates the length of the delimiter for a given message size
-func DelimLen(x uint64) int {
-	return 8 - bits.LeadingZeros64(x)%8
+func DelimLen(size uint64) int {
+	lenBuf := make([]byte, binary.MaxVarintLen64)
+	return binary.PutUvarint(lenBuf, size)
```
Comment (evan-forbes) [inline pkg/shares/utils.go:14]: reverted this change to calculate the delim len how we used to, as the new way wasn't doing what we thought. We still need to add a unit test to make sure we don't make the same mistake again. This is implicitly tested in the new fuzzer added in this PR, and message inclusion checks.

File: app/process_proposal.go [hunk 3]
```diff
@@ -122,7 +61,70 @@ func (app *App) ProcessProposal(req abci.RequestProcessProposal) abci.ResponsePr
...
 
 	if !bytes.Equal(dah.Hash(), req.Header.DataHash) {
-		logInvalidPropBlockError(app.Logger(), req.Header, "proposed data root differs from calculated data root", err)
+		logInvalidPropBlock(app.Logger(), req.Header, "proposed data root differs from calculated data root")
+		return abci.ResponseProcessProposal{
+			Result: abci.ResponseProcessProposal_REJECT,
+		}
+	}
+
+	// iterate over all of the MsgPayForData transactions and ensure that they
```
Comment (rootulp) [inline app/process_proposal.go:70]:
```suggestion
	// iterate over all of the MsgPayForData transactions and ensure that their
```

**Review by rootulp — DISMISSED** [review]

**Review by rootulp — APPROVED** [review]
